src/Components/WebAssembly/WebAssembly/src/Hosting/MutableConfiguration.cs

```diff
@@ -0,0 +1,197 @@
+        private readonly object mutex = new object();
+
+        /// <summary>
+        /// Returns the sources used to obtain configuration values.
+        /// </summary>
+        public IList<IConfigurationSource> Sources => _sources;
+
+        /// <summary>
+        /// Returns the providers user to obtain configuration values.
+        /// </summary>
+        public IEnumerable<IConfigurationProvider> Providers => _providers;
+
+        /// <summary>
+        /// Gets a key/value collection that can be used to share data between the <see cref="IConfigurationBuilder"/>
+        /// and the registered <see cref="IConfigurationProvider"/>s.
+        ///
+        /// In this implementation, this largely exists as a way to satisfy the
+        /// requirements of the IConfigurationBuilder and is not populated by
+        /// <see cref="MutableConfiguration"/> with any meaningful info.
+        /// </summary>
+        public IDictionary<string, object> Properties { get; } = new Dictionary<string, object>();
+
+        /// <summary>
+        /// Gets or sets a configuration value in all providers that are currently
+        /// registered on the <see cref="MutableConfiguration"/> object.
+        /// </summary>
+        /// <param name="key">The configuration key.</param>
+        /// <returns>The configuration value.</returns>
+        public string this[string key] {
+            get {
+                lock (mutex)
+                {
+                    for (var i = _providers.Count - 1; i >= 0; i--)
+                    {
+                        var provider = _providers[i];
+
+                        if (provider.TryGet(key, out var value))
+                        {
+                            return value;
+                        }
+                    }
+                }
+                return null;
+            }
+            set {
+                if (_providers.Count == 0)
+                {
+                    throw new InvalidOperationException();
+                }
+
+                lock (mutex)
+                {
+                    foreach (var provider in _providers)
+                    {
+                        provider.Set(key, value);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets a configuration sub-section with the specified key.
+        /// </summary>
+        /// <param name="key">The key of the configuration section.</param>
+        /// <returns>The <see cref="IConfigurationSection"/>.</returns>
+        /// <remarks>
+        ///     This method will never return <c>null</c>. If no matching sub-section is found with the specified key,
+        ///     an empty <see cref="IConfigurationSection"/> will be returned.
+        /// </remarks>
+        public IConfigurationSection GetSection(string key) => new ConfigurationSection(this, key);
+
+        public IEnumerable<IConfigurationSection> GetChildren() => GetChildrenImplementation(null);
```

Comment:
Does it make sense to create an `IConfigurationRoot` object like build does (hold on to it), and then always delegate these implementations to it?

Reply:
I had started off with this model but hit a snag at this point and ended up with the approach you see here. I can't remember what the particular snag was but I'll refactor to that model and see if I can recall it.